In [1]:
from __future__ import unicode_literals

import pandas as pd
import numpy as np
from sklearn import preprocessing
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split

# 1. Preparing data - Latest 12 weeks of life of the banana plant
---

In [2]:
temperature_l12w = pd.read_csv('../../../../data/raw/FincaPorvenir/Metereologico/Latest_12-weeks_May-4_August-03_2018/' \
                                 'Temperatura_May-4_August-03_2018.csv', )

In [3]:
print(temperature_l12w.shape)
temperature_l12w.head()

(4398, 2)


Fecha:  Temperatura (C)
0  2018-05-04 00:07:38            23.75
1  2018-05-04 00:37:39            23.68
2  2018-05-04 01:07:38            23.68
3  2018-05-04 01:37:39            23.56
4  2018-05-04 02:07:39            23.50

## 1.1. We evaluate if this dataset has null type `NaN`
---

In [4]:
print(temperature_l12w.isnull().any())
temperature_l12w.isnull().values.any()

Fecha:             False
Temperatura (C)    False
dtype: bool


False

Don't have null values

## 1.2. Selecting  relevant index columns features
---

Since the dataset has a column called **`Fecha:`**, which is not a numerical value,

it will be removed so that it does not interfere **with our subsequent scaling**, 

so we are only going to reference the values or samples of the column

** `Temperatura (C)` ** and assigning them to the matrix 

`temperature_l12w_array` created such as follow:

In [5]:
temperature_l12w_array = temperature_l12w.iloc[:,1].values

In [6]:
# Rehape the luminosity_luxes array 
temperature_l12w_array = temperature_l12w_array.reshape(-1, 1)
temperature_l12w_array

array([[23.75],
       [23.68],
       [23.68],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

## 1.3 Generating descriptive data to dataset
---

In [7]:
col = ['Temperatura (C)']
temperature_l12w_values_df = pd.DataFrame(temperature_l12w_array, columns=col)
temperature_l12w_values_df_describe = temperature_l12w_values_df.describe()

In [8]:
temperature_l12w_values_df_describe

Temperatura (C)
count      4398.000000
mean         27.531389
std           4.273432
min          21.370000
25%          24.000000
50%          25.900000
75%          31.060000
max          40.250000

In [9]:
# Export this descriptive data to comma separated values and java script object notation
temperature_l12w_values_df_describe.to_csv('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                              'Temperature_Describe_May-4_August-03_2018.cvs', sep=',', header=True, index=True)
temperature_l12w_values_df_describe.to_json('../../../../data/interim/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n'
                                     'Temperature_Describe_May-4_August-03_2018.json')

# 2. Creating Temperature Training and Testing datasets

We have a **`temperature_l12w_values_df`** dataset with 4398 samples rows.

In [11]:
print(temperature_l12w_values_df.shape)
temperature_l12w_values_df.head()

(4398, 1)


Temperatura (C)
0            23.75
1            23.68
2            23.68
3            23.56
4            23.50

We'll divide it into two differents datasets:

- Training dataset
- Testing dataset

This is executed through [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function of this way:

`train_test_split` receive as a data parameter a numpy array, we have to turn the 

**`temperature_l12w_values_df`** dataframe to numpy array such as follow:

In [13]:
# numpy_temperature_values = temperature_values_df.reset_index().values
numpy_temperature_l12w_values = temperature_l12w_values_df.values

In [14]:
numpy_temperature_l12w_values

array([[23.75],
       [23.68],
       [23.68],
       ...,
       [23.81],
       [23.5 ],
       [23.62]])

We compose the following datasets from **`numpy_temperature_l12w_values`** array :

- `temperature_l12w_values_train`, which is the training matrix
- `temperature_l12w_values_test`, the testing matrix

We use the [train_test_split](http://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html  "sklearn.model_selection.train_test_split") function to create the training and testing dataset. Their parameters are:

![alt text](https://cldup.com/hukPWvvLxt-3000x3000.png "klearn.model_selection.train_test_split")


- The first parameter **should be an array**, then we pass the **`numpy_temperature_l12w_values`** which contain all column features of luminosity.

Of this way, we pass all data (4390 samples rows) to from them `temperature_l12w_values_train`

and `temperature_l12w_values_test` will be created

- The `test_size=0.5` parameter means a 50% division; which means 
that half of the data goes to the test dataset and the other half 
goes to the training dataset.

A good choice for the size of tests **is usually 0.2 ie 20% or 0.25 or even 30%.** 

In some rare cases we will have 40% but almost never 0.5 or 50%

We choose 20% which means that we will have 20% of 9752 samples or observations for the test data set, 

in this case **4390 * 0.2 = 878 samples or records for the test data set**


- The `train_size` parameter is the training dataset size.  **`test_size + train_size = 1 or 100%`**, 
then isn't necessary include it, because wheter we include to `test_size = 0.2`, then the remaining 
data will be to `train_size` this means **0.8 or 80%**

This means that **4390 * 0.8 = 3512  training dataset samples rows. **


- random_state is a seed or data source for generating random values for the data sets. 

If this parameter is not passed, the data will be generated in a random way, but in the way as by default numpy works them.

### 2.1  Creating Training and testing temperature datasets
---

- `temperature_l12w_values_train`, which is the training matrix
- `temperature_l12w_values_test`, the testing matrix

In [15]:
temperature_l12w_values_train, temperature_l12w_values_test = train_test_split(numpy_temperature_l12w_values, test_size = 0.2)

In [17]:
# We have 3518 rows to luminosity_luxes_train
print(type(temperature_l12w_values_train))
print("The dimensionality of wind_direction training dataset is: " +'\n' , temperature_l12w_values_train.shape)
print('\n')

# And we have 880 rows to luminosity_luxes_test
print(type(temperature_l12w_values_test))
print("The dimensionality of wind_direction testing dataset is: " +'\n' , temperature_l12w_values_test.shape)

<class 'numpy.ndarray'>
The dimensionality of wind_direction training dataset is: 
 (3518, 1)


<class 'numpy.ndarray'>
The dimensionality of wind_direction testing dataset is: 
 (880, 1)


In this way, the model, as you progressively learn the correlations in the training set, the better the prediction of the results in the test set.

But if the model learns too much from memory the correlations of the training sets,

that is to say, when one learns from memory and does not understand things, then it

will have problems to predict what is happening on the set of tests, because it is

learned for difficult correlations, if the logic is not well understood and you can

not make good predictions. This is called overfitting or overfitting

The really important thing is to understand that we need to have two different datasets

- Training set with which the ML model learns
- Test set, on which we test whether the ML model correctly learned the correlations


---
##  3. Feature Scaling Temperature training and testing dataset  
---

[This article post](http://benalexkeen.com/feature-scaling-with-scikit-learn/  "Feature Scaling with scikit-learn") it's a great reference to explore the features scaling methods
on scikit learn

- `StandardScaler` assume that data is normally distributed at the level of each characteristic or variable. If the data is not normally distributed, it is not the best alternative to use for scaling.  

- `Min-Max Scaler` it is probably the most famous scaling algorithm and what it does is resize the range to leave it in a dimension of 0 to 1 or -1 and 1 (in case there are negative values in the original dataset of input data)

This scale of maximums and minimums works best for cases where standard scaling may not work properly. If the distribution is not Gaussian or the standard deviation is very small the escalation of maximums and minimums is the best idea.
However, it is sensitive to outliers or outliers, so if there are outliers in the data it is better to consider robust scaling.

- `Robust Scaler` it is similar to the previous one of maximums and minimums, only that it uses interquartile ranges instead of maximums and minimums, which makes it robust for the outliers

- `Normalizer` which scales each value, dividing each value by its magnitude in n dimensional spaces for n number of characteristics.


---

We will use the scaling of maximums and minimums to scale the water table data, because the standard deviation is very small, 

it does not have atypical values and it does not follow a normal distribution (you have to check this)

We apply the maximum and minimum scaling. We provide a rank or base scale that **will be between 0 and 1** using an object,

[MinMaxScaler](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html "sklearn.preprocessing.MinMaxScaler") which transforms each characteristic, (in this 

case it will be the columns feature of `Temperature (C)`) individually according to a given range.

Product of its applicability, generates these attributes in the dataset, already transformed: 

![alt text](https://cldup.com/lTIv4HXgTk-3000x3000.png "sklearn.preprocessing.MinMaxScaler")

### 3.1 We apply maximium and minimum feature scaling to Temperature training dataset

In [19]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature training data " + '\n', temperature_l12w_values_train)

Remember our temperature training data 
 [[24.93]
 [29.37]
 [25.43]
 ...
 [23.68]
 [26.81]
 [34.81]]


With the [fit](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.fit  "MinMaxScaler.fit")
we compute the maximum and minimum value of  `temperature_l12w_values_train` dataset to be used in the subsequent scaling 

We assing these values to `minmax_scale_training` variable.

In [20]:
minmax_scale_training = scaler.fit(temperature_l12w_values_train.astype(float))
# print(minmax_scale_training.data_max_)
# http://terrapinssky.blogspot.com/2017/10/pythonresolved-dataconversionwarning.html

Then, we apply the [transform](http://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html#sklearn.preprocessing.MinMaxScaler.transform "MinMaxScaler.transform") method to transform these data to maximum and mínimum scale value. 

Here, with this process, the `temperature_l12w_values_train` data are scaled between **0 to 1**  selected range 

In [21]:
# transform precipitations_train data to maximum and mínimum scale value. 
temperature_l12w_minmax_training = minmax_scale_training.transform(temperature_l12w_values_train)

In [22]:
print ("And, these are our scaled data: " + '\n')
temperature_l12w_minmax_training

And, these are our scaled data: 



array([[0.18855932],
       [0.42372881],
       [0.21504237],
       ...,
       [0.12235169],
       [0.28813559],
       [0.71186441]])

In [23]:
print('Temperature Training dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_training[:,0].min()))

print('Temperature Training dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_training[:,0].max()))

Temperature Training dataset. Minimum value after MaxMinScaler:
Temperature=0.0
Temperature Training dataset. Maximum value after MaxMinScaler:
Temperature=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_l12w_minmax_training` numpy array

- We export this array to comma separated values 


In [27]:
temperature_l12w_train_df = pd.DataFrame(temperature_l12w_minmax_training, columns=col)

In [28]:
temperature_l12w_train_df

Temperatura (C)
0            0.188559
1            0.423729
2            0.215042
3            0.589513
4            0.073093
5            0.069386
6            0.148835
7            0.079449
8            0.122352
9            0.307733
10           0.510064
11           0.241525
12           0.486758
13           0.201801
14           0.188559
15           0.152542
16           0.466631
17           0.288136
18           0.261653
19           0.294492
20           0.129237
21           0.182203
22           0.387182
23           0.238347
24           0.519597
25           0.576271
26           0.132415
27           0.288136
28           0.109110
29           0.152542
...               ...
3488         0.129237
3489         0.529661
3490         0.195445
3491         0.592691
3492         0.658898
3493         0.066208
3494         0.827860
3495         0.426907
3496         0.115996
3497         0.142479
3498         0.691737
3499         0.102754
3500         0.642479
3501         0.231992
3502         0.142479
3503         0.278072
3504         0.198623
3505         0.337924
3506         0.076271
3507         0.168962
3508         0.162076
3509         0.701801
3510         0.112818
3511         0.221928
3512         0.099576
3513         0.493114
3514         0.331038
3515         0.122352
3516         0.288136
3517         0.711864

[3518 rows x 1 columns]

In this way we have the dataset `temperature_l12w_train_df` standardized and training, and export it to a .csv file

In [29]:
temperature_l12w_train_df.to_csv('../../../../data/processed/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TRAINING_May-4_August-03_2018.csv', sep=',', header=True, index=False)

### 3.2 We apply maximium and minimum feature scaling to Temperature testing dataset


In [30]:
# We provide a base scale range
scaler = MinMaxScaler(feature_range=(0, 1))

print("Remember our temperature testing data " + '\n', temperature_l12w_values_test)

Remember our temperature testing data 
 [[24.43]
 [32.68]
 [23.31]
 [34.  ]
 [26.12]
 [23.12]
 [23.68]
 [27.12]
 [23.81]
 [30.25]
 [33.25]
 [34.62]
 [35.43]
 [23.  ]
 [39.12]
 [23.31]
 [24.56]
 [23.5 ]
 [29.18]
 [34.  ]
 [30.75]
 [23.5 ]
 [24.68]
 [23.37]
 [35.37]
 [35.62]
 [33.75]
 [27.93]
 [35.68]
 [32.62]
 [35.62]
 [26.  ]
 [23.62]
 [32.31]
 [23.06]
 [24.87]
 [33.37]
 [29.68]
 [22.87]
 [23.56]
 [32.56]
 [23.12]
 [24.87]
 [24.37]
 [25.18]
 [32.18]
 [22.93]
 [23.56]
 [31.  ]
 [25.37]
 [27.87]
 [29.06]
 [35.56]
 [24.12]
 [26.  ]
 [25.81]
 [23.37]
 [24.18]
 [23.25]
 [27.93]
 [23.62]
 [24.56]
 [25.12]
 [24.62]
 [34.62]
 [24.62]
 [27.62]
 [25.06]
 [23.  ]
 [25.25]
 [24.12]
 [35.06]
 [24.5 ]
 [31.25]
 [33.75]
 [23.25]
 [24.37]
 [26.31]
 [25.37]
 [24.  ]
 [32.25]
 [24.  ]
 [37.06]
 [23.37]
 [29.  ]
 [23.37]
 [34.  ]
 [28.06]
 [34.12]
 [23.62]
 [23.43]
 [25.25]
 [29.25]
 [34.87]
 [30.93]
 [30.25]
 [24.87]
 [32.12]
 [28.62]
 [22.87]
 [22.56]
 [24.  ]
 [31.81]
 [24.5 ]
 [23.18]
 [23.12]
 [35.6

In [31]:
minmax_scale_test = scaler.fit(temperature_l12w_values_test.astype(float))
# transform luminosity_luxes_test data to maximum and mínimum scale value. 
temperature_l12w_minmax_test = minmax_scale_test.transform(temperature_l12w_values_test)

In [32]:
print ("And, these are our testing scaled data: " + '\n')
temperature_l12w_minmax_test

And, these are our testing scaled data: 



array([[0.16666667],
       [0.625     ],
       [0.10444444],
       [0.69833333],
       [0.26055556],
       [0.09388889],
       [0.125     ],
       [0.31611111],
       [0.13222222],
       [0.49      ],
       [0.65666667],
       [0.73277778],
       [0.77777778],
       [0.08722222],
       [0.98277778],
       [0.10444444],
       [0.17388889],
       [0.115     ],
       [0.43055556],
       [0.69833333],
       [0.51777778],
       [0.115     ],
       [0.18055556],
       [0.10777778],
       [0.77444444],
       [0.78833333],
       [0.68444444],
       [0.36111111],
       [0.79166667],
       [0.62166667],
       [0.78833333],
       [0.25388889],
       [0.12166667],
       [0.60444444],
       [0.09055556],
       [0.19111111],
       [0.66333333],
       [0.45833333],
       [0.08      ],
       [0.11833333],
       [0.61833333],
       [0.09388889],
       [0.19111111],
       [0.16333333],
       [0.20833333],
       [0.59722222],
       [0.08333333],
       [0.118

In [33]:
print('Temperature Testing dataset. Minimum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_test[:,0].min()))

print('Temperature Testing dataset. Maximum value after MaxMinScaler:\nTemperature={:.1f}'
      .format(temperature_l12w_minmax_test[:,0].max()))

Temperature Testing dataset. Minimum value after MaxMinScaler:
Temperature=0.0
Temperature Testing dataset. Maximum value after MaxMinScaler:
Temperature=1.0


Then, our  MinMaxScaler normalized training dataset is `temperature_l12w_minmax_test` numpy array

- We export this array to comma separated values

In [35]:
temperature_l12w_test_df = pd.DataFrame(temperature_l12w_minmax_test, columns=col)

In [36]:
print(temperature_l12w_test_df.shape)
temperature_l12w_test_df

(880, 1)


Temperatura (C)
0           0.166667
1           0.625000
2           0.104444
3           0.698333
4           0.260556
5           0.093889
6           0.125000
7           0.316111
8           0.132222
9           0.490000
10          0.656667
11          0.732778
12          0.777778
13          0.087222
14          0.982778
15          0.104444
16          0.173889
17          0.115000
18          0.430556
19          0.698333
20          0.517778
21          0.115000
22          0.180556
23          0.107778
24          0.774444
25          0.788333
26          0.684444
27          0.361111
28          0.791667
29          0.621667
..               ...
850         0.226111
851         0.240000
852         0.423889
853         0.326667
854         0.062778
855         0.173889
856         0.420556
857         0.240000
858         0.087222
859         0.326667
860         0.448333
861         0.882222
862         0.149444
863         0.180556
864         0.152778
865         0.361111
866         0.302222
867         0.132222
868         0.107778
869         0.160000
870         0.125000
871         0.104444
872         0.736111
873         0.277778
874         0.118333
875         0.194444
876         0.083333
877         0.083333
878         0.146111
879         0.236111

[880 rows x 1 columns]

In [38]:
temperature_l12w_test_df.to_csv('../../../../data/processed/Temperature/Latest_12-weeks_May-4_August-03_2018/' +'\n' 
                                 'Temperature_Normalized_TESTING_May-4_August-03_2018.csv', sep=',', header=True, index=False)